sync/src/transaction_queue.rs

```diff
@@ -0,0 +1,622 @@
+// Copyright 2015, 2016 Ethcore (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+// TODO [todr] - own transactions should have higher priority
+
+//! Transaction Queue
+
+use std::vec::Vec;
+use std::cmp::{Ordering};
+use std::collections::{HashMap, BTreeSet};
+use util::uint::{Uint, U256};
+use util::hash::{Address};
+use util::table::*;
+use ethcore::transaction::*;
+
+
+#[derive(Clone, Debug)]
+struct VerifiedTransaction {
+	tx: SignedTransaction,
+	nonce_height: U256
+}
+
+impl VerifiedTransaction {
+	pub fn new(tx: SignedTransaction, nonce_height: U256) -> VerifiedTransaction {
+		VerifiedTransaction {
+			tx: tx,
+			nonce_height: nonce_height
+		}
+	}
+
+	pub fn sender(&self) -> Address {
+		self.tx.sender().unwrap()
+	}
+}
+
+impl Eq for VerifiedTransaction {}
+impl PartialEq for VerifiedTransaction {
+	fn eq(&self, other: &VerifiedTransaction) -> bool {
+		self.cmp(other) == Ordering::Equal
+	}
+}
+impl PartialOrd for VerifiedTransaction {
+	fn partial_cmp(&self, other: &VerifiedTransaction) -> Option<Ordering> {
+		Some(self.cmp(other))
+	}
+}
+impl Ord for VerifiedTransaction {
+	fn cmp(&self, b: &VerifiedTransaction) -> Ordering {
+		// First check nonce_height
+		if self.nonce_height != b.nonce_height {
+			return self.nonce_height.cmp(&b.nonce_height);
+		}
+
+		// Then compare gas_prices
+		let a_gas = self.tx.gas_price;
+		let b_gas = b.tx.gas_price;
+		if a_gas != b_gas {
+			return a_gas.cmp(&b_gas);
+		}
+
+		// Compare nonce
+		let a_nonce = self.tx.nonce;
+		let b_nonce = b.tx.nonce;
+		if a_nonce != b_nonce {
+			return a_nonce.cmp(&b_nonce);
+		}
+
+		// and senders
+		let a_sender = self.sender();
+		let b_sender = b.sender();
+		a_sender.cmp(&b_sender)
+	}
+}
+
+struct TransactionsByPriorityAndAddress {
+	priority: BTreeSet<VerifiedTransaction>,
+	address: Table<Address, U256, VerifiedTransaction>,
```

Comment:
I suggest the following to avoid duplicating transaction data:
Use an additional `HashMap<H256, Verified Transaction>` to store the actual transaction data.
The key to `priority` can be a small struct that only contains nonce height and gas price, and the value is transaction hash
`address` should hold a hash as a value too.